Implement proper balance transfer in task pallet and minor refactoring.

vivekvpandya opened the pull request.

vivekvpandya committed:
Implement proper balance transfer in task pallet and minor refactoring.

stojanov-igor commented on pallets/task/src/lib.rs:
```diff
@@ -298,6 +302,11 @@ pub mod pallet {
 
+			let task = Self::tasks(&task_id).ok_or(<Error<T>>::TaskNotExist)?;
+			// pay volunteer
+			let sub_account = Self::account_id(&task_id);
```
We are creating a separate unique account based on taskID to serve as an intermediary for a balance transfer?

Can this account be controlled?

vivekvpandya replied:
So each task_id has unique sub_account from pallet, so it is secure (I feel this is same as PDAs in Solana). But if you found anything wrong with this I am okay for other options.

vivekvpandya replied:
So it is not exactly like PDAs in solana but to me it seems quite safe as it is based on Task::Hash which is unique (as guaranteed by Hashing algorithms) see how sub_account is created at here https://substrate.stackexchange.com/questions/477/how-can-we-get-a-pallets-sub-account-address-using-polkadotjs-api

stojanov-igor commented on pallets/task/src/tests.rs:
```diff
@@ -519,18 +518,6 @@ fn task_can_be_rejected_by_creator(){
 }
 
-#[test]
-fn transfer_balance_works(){
```
Please add new tests!

vivekvpandya replied:
That test just test same thing which has been well tested in pallet_balances, or pallet_assets or orml_token etc.
So I don't thing we need that.

stojanov-igor reviewed (changes requested):
Add unit tests! 

The rest seems fine.

vivekvpandya commented:
> Add unit tests!
> 
> The rest seems fine.

I will update tests to check balances after creation/updation and deletion of tasks.

vivekvpandya committed:
Update tasks tests to check balance updates after create_task, update_task and accept_task.
Use u128 for AcccountId associated type in system::Config otherwise get_sub_account() can not produce unique value.

vivekvpandya commented:
ping for review @JosephKnecht-lab

stojanov-igor commented on pallets/task/src/tests.rs:
```diff
@@ -1,5 +1,6 @@
 use crate::TaskStatus;
 use crate::{mock::*, Error};
+use frame_support::traits::fungible::Inspect;
```
why you need this? Its not used...

vivekvpandya replied:
it provides balance() interface for pallet-balances , I have used it to test balance.

stojanov-igor commented on pallets/task/src/tests.rs:
```diff
@@ -41,6 +42,9 @@ fn create_new_task(){
 
 		// Ensure new task can be created with [signer, specification, budget, deadline]
 		assert_ok!(Task::create_task(Origin::signed(1), TITLE.to_vec(), vec, 7, get_deadline()));
+		assert_eq!(Balances::balance(&1), 993);
```
We should keep the tests small. Instead of adding multiple asserts into existing tests, please create 2-3 new tests that are only testing for balances when:

- Balance is transferred to the escrow account successfully when the task is started.
- Balance is transferred to volunteer once the task is accepted.
- Possible errors that might be thrown. 

The tests you have here are good, they just need to be specified in their own unit tests.

vivekvpandya replied:
Alright!

vivekvpandya commented:
ping for review!

stojanov-igor reviewed (approved):
Looks better
